Core/clim-basic/ports.lisp

```diff
@@ -171,107 +176,119 @@
+        (dispatch-event-copy s event 'pointer-enter-event)))))
+
+(defmethod distribute-event ((port basic-port) event)
+  (dispatch-event (event-sheet event) event))
+
+;;; In the most general case we can't tell whenever all sheets are
+;;; mirrored or not. So a default method for pointer-events operates
+;;; under the assumption that we must deliver events to sheets which
+;;; doesn't have a mirror and that the sheet grabbing and pressing is
+;;; not implemented by the port (we emulate it). -- jd 2019-08-21
 (defmethod distribute-event ((port basic-port) (event pointer-event))
-  (when-let ((sheet (port-pointer-sheet port)))
-    (if (eq sheet (event-sheet event))
-        (call-next-method)
-        ;; events are immutable (explicitly stated in the spec) - that's why we
-        ;; need to make an event copy for single-mirrored sheets - event-sheet
-        ;; is not the same as sheet we want to distribute event to.
-        (let ((new-event (shallow-copy-object event)))
-          (setf (slot-value new-event 'sheet) sheet)
-          (unless (eq (sheet-mirrored-ancestor sheet)
-                      (sheet-mirrored-ancestor (event-sheet event)))
-            (multiple-value-bind (cx cy)
-                (untransform-position (sheet-delta-transformation (sheet-mirrored-ancestor sheet) nil)
-                                      (slot-value new-event 'graft-x)
-                                      (slot-value new-event 'graft-y))
-              (setf (slot-value new-event 'x) cx
-                    (slot-value new-event 'y) cy)))
-          (dispatch-event sheet new-event)))))
+  ;; When we receive pointer event we need to take into account
+  ;; unmirrored sheets and grabbed/pressed sheets.
+  ;;
+  ;; - Grabbed sheet steals all pointer events (non-local exit)
+  ;; - Pressed sheet receives replicated motion events
+  ;; - Pressing/releasing the button assigns pressed-sheet
+  ;; - Pointer motion may result in synthetized boundary events
+  ;; - Events are delivered to the innermost child of the sheet
+  (let ((grabbed-sheet (port-grabbed-sheet port))
+        (pressed-sheet (port-pressed-sheet port))
+        (old-pointer-sheet (port-pointer-sheet port))
+        (new-pointer-sheet (get-pointer-event-sheet event)))
+    (when grabbed-sheet
+      (unless (typep event '(or pointer-enter-event pointer-exit-event))
+        (dispatch-event-copy grabbed-sheet event))
+      (return-from distribute-event))
+    ;; Synthetize boundary change events and update the port.
+    (synthetize-enter-exit-events old-pointer-sheet new-pointer-sheet event)
+    (setf (port-pointer-sheet port) new-pointer-sheet)
+    ;; Set the pointer cursor.
+    (when-let ((cursor-sheet (or pressed-sheet new-pointer-sheet)))
+      (let ((old-pointer-cursor (port-lookup-current-pointer-cursor port (event-sheet event)))
+            (new-pointer-cursor (sheet-pointer-cursor new-pointer-sheet)))
+	(unless (eql old-pointer-cursor new-pointer-cursor)
```

Comment:
Indentation uses tabs in these two lines.

Reply:
fixed